### PR TITLE
Allow index files to be provided

### PR DESCRIPTION
**Pull request by SusiJo**
Many thanks to contributing to qbic-pipelines/bamtofastq!

Please fill in the appropriate checklist below (delete whatever is not relevant). These are the most common things requested on pull requests (PRs).

## PR checklist
 - [ ] This comment contains a description of changes (with reason)
 - [ ] If you've fixed a bug or added code that should be tested, add tests!
 - [ ] If necessary, add test data to `testdata/` 
 - [ ] Ensure the test suite passes (`nextflow run . -profile test,docker`).
 - [ ] Make sure your code lints (`nf-core lint .`). 
 - [ ] Documentation in `docs` is updated
 - [ ] `CHANGELOG.md` is updated
 - [ ] `README.md` is updated

**Learn more about contributing:** https://github.com/qbic-pipelines/bamtofastq/tree/master/.github/CONTRIBUTING.md

### CODE REVIEW

File: main.nf [hunk 1]
```diff
@@ -92,13 +94,30 @@ if(params.input && !params.chr) { //Checks whether bam file(s) and no chromosome
...
+	Channel
+	.fromPath(params.index_files, checkIfExists:true)	
+	.map{file -> tuple(file.name.replaceAll(".bai", ''), file)}	
+	.into { bai_files_idxstats;
+			bai_files_chr}
+	
+}
+else {
+	bai_files_idxstats = Channel.empty();
+	bai_files_chr = Channel.empty();
```
Comment (ggabernet) [inline main.nf:119]:
```suggestion
	bai_files_idxstats = Channel.empty()
	bai_files_chr = Channel.empty()
```

File: main.nf [hunk 2]
```diff
@@ -180,29 +200,114 @@ process get_software_versions {
...
+		                
+
+	    script:
+	    //If multiple chr were specified, then join space separated list for naming: chr1 chr2 -> chr1_chr2, also resolve region specification with format chr:start-end
+	    chr_list_joined = params.chr.split(' |-|:').size() > 1 ? params.chr.split(' |-|:').join('_') : params.chr
+	    """    
+	    samtools view -hb $bam ${params.chr} -@$task.cpus -o ${name}.${chr_list_joined}.bam
+	    """
+	 	}
+	 }
```
Comment (ggabernet) [inline main.nf:246]:
```suggestion
	 }
    }
```

File: main.nf [hunk 3]
```diff
@@ -180,29 +200,114 @@ process get_software_versions {
...
+		set val(name), file(bai) from ch_index_files
+				
+		output:
+		file "*.idxstats" into ch_bam_idxstat_mqc		
+		script:
+		"""				
+		samtools idxstats $bam > ${bam}.idxstats
+		"""
+
+		}
```
Comment (ggabernet) [inline main.nf:263]:
```suggestion
	}
```

File: main.nf [hunk 4]
```diff
@@ -180,29 +200,114 @@ process get_software_versions {
...
+}
```
Comment (ggabernet) [inline main.nf:264]:
```suggestion
} else {
```

File: main.nf [hunk 5]
```diff
@@ -180,29 +200,114 @@ process get_software_versions {
...
 
-    script:
-    //If multiple chr were specified, then join space separated list for naming: chr1 chr2 -> chr1_chr2, also resolve region specification with format chr:start-end
-    chr_list_joined = params.chr.split(' |-|:').size() > 1 ? params.chr.split(' |-|:').join('_') : params.chr
-    """
-    samtools index -@$task.cpus $bam 
-    samtools view -hb $bam ${params.chr} -@$task.cpus -o ${name}.${chr_list_joined}.bam
-    """
-  }
+else {
```
Comment (ggabernet) [inline main.nf:270]:
```suggestion
```

File: main.nf [hunk 6]
```diff
@@ -180,29 +200,114 @@ process get_software_versions {
...
+	    set val("${name}.${chr_list_joined}"), file("${name}.${chr_list_joined}.bam") into bam_files_check
+
+	    script:
+	    //If multiple chr were specified, then join space separated list for naming: chr1 chr2 -> chr1_chr2, also resolve region specification with format chr:start-end
+	    chr_list_joined = params.chr.split(' |-|:').size() > 1 ? params.chr.split(' |-|:').join('_') : params.chr
+	    """    
+	    samtools view -hb $bam ${params.chr} -@$task.cpus -o ${name}.${chr_list_joined}.bam
+	    """
+	  	}
+	  }
```
Comment (ggabernet) [inline main.nf:291]:
```suggestion
	  }
    }
```

**Comment by ggabernet** [comment]
Hi @SusiJo ,

thanks for the PR, I've added some comments. I did not see a clear reason so far for the failed tests but will investigate further.

**Comment by ggabernet** [comment]
@FriederikeHanssen could you also have a look at the PR whenever you have some time?

File: main.nf [hunk 7]
```diff
@@ -29,7 +29,8 @@ def helpMessage() {
     Other options:   
       --outdir                 [file]  The output directory where the results will be saved
       --chr                     [str]  Only use reads mapping to a specific chromosome/region. Has to be specified as in bam: i.e chr1, chr{1..22} (gets all reads mapping to chr1 to 22), 1, "X Y", incorrect naming will lead to a potentially silent error
-      --no_read_QC             [bool]  If specified, no quality control will be performed on extracted reads. Useful, if this is done anyways in the subsequent workflow
+      --index_files            [file]  Path to bai index files
+	  --no_read_QC             [bool]  If specified, no quality control will be performed on extracted reads. Useful, if this is done anyways in the subsequent workflow
```
Comment (FriederikeHanssen) [inline main.nf:33]:
```suggestion
     --no_read_QC             [bool]  If specified, no quality control will be performed on extracted reads. Useful, if this is done anyways in the subsequent workflow
```

File: main.nf [hunk 8]
```diff
@@ -82,6 +83,7 @@ if(params.input && !params.chr) { //Checks whether bam file(s) and no chromosome
         .map { file -> tuple(file.name.replaceAll(".bam",''), file) } // map bam file name w/o bam to file 
         .into { bam_files_check; 
                 bam_files_flagstats;
+				bam_files_index;
```
Comment (FriederikeHanssen) [inline main.nf:86]:
```suggestion
		bam_files_index;
```

File: main.nf [hunk 9]
```diff
@@ -92,13 +94,30 @@ if(params.input && !params.chr) { //Checks whether bam file(s) and no chromosome
         .map { file -> tuple(file.name.replaceAll(".bam",''), file) } // map bam file name w/o bam to file 
         .into { bam_chr;
                 bam_files_flagstats;
+				bam_files_index;
```
Comment (FriederikeHanssen) [inline main.nf:97]:
```suggestion
		bam_files_index;
```

File: main.nf [hunk 10]
```diff
@@ -92,13 +94,30 @@ if(params.input && !params.chr) { //Checks whether bam file(s) and no chromosome
...
+/*
+ * Create a channel for input bai files
+ */
+
+if (params.index_files) {
+	Channel
+	.fromPath(params.index_files, checkIfExists:true)	
+	.map{file -> tuple(file.name.replaceAll(".bai", ''), file)}	
+	.into { bai_files_idxstats;
+			bai_files_chr}
```
Comment (FriederikeHanssen) [inline main.nf:114]:
```suggestion
		bai_files_chr}
```

Comment (FriederikeHanssen) [inline main.nf:114]:
```suggestion
	  bai_files_chr}
```

File: main.nf [hunk 11]
```diff
@@ -110,6 +129,7 @@ summary['Max Resources']                                              = "$params
 if (workflow.containerEngine) summary['Container']                    = "$workflow.containerEngine - $workflow.container"
 summary['Output dir']                                                 = params.outdir
 if (params.chr) summary['Only reads mapped to chr']                   = params.chr
+if (params.index_files) summary['Index files available']			  = params.index_files
```
Comment (FriederikeHanssen) [inline main.nf:132]:
```suggestion
if (params.index_files) summary['Index files available']		= params.index_files
```

File: main.nf [hunk 12]
```diff
@@ -180,29 +200,114 @@ process get_software_versions {
...
+else {	
+	// Extract reads mapping to specific chromosome(s)
+	if (params.chr){
+	  process extractReadsMappingToChromosomeBAI{
+	    tag "${name}.${chr_list_joined}"
+	    label 'process_medium'
+  
+	    input:
+	    set val(name), file(bam) from bam_chr 
+		set val(name), file(bai) from bai_files_chr
```
Comment (FriederikeHanssen) [inline main.nf:279]:
```suggestion
	set val(name), file(bai) from bai_files_chr
```

**Review by FriederikeHanssen — COMMENTED** [review]
Can you maybe elaborate on why you choose to copy the process for the chromosomal read extraction? Would it not be possible to simplify the logic, since they both take the same input and output? Like so:

Channel index = Channel.empty()
if (no_input_index){
       Process: IndexInputBam
     index = IndexOutputBam.out
}else {
    index = input_indices
}

if (params.chr) {
 process: extractReadsmappedtoChromosome(bam, index)
}


Could possibly be compressed further.

File: main.nf [hunk 13]
```diff
@@ -29,7 +29,8 @@ def helpMessage() {
     Other options:   
       --outdir                 [file]  The output directory where the results will be saved
       --chr                     [str]  Only use reads mapping to a specific chromosome/region. Has to be specified as in bam: i.e chr1, chr{1..22} (gets all reads mapping to chr1 to 22), 1, "X Y", incorrect naming will lead to a potentially silent error
-      --no_read_QC             [bool]  If specified, no quality control will be performed on extracted reads. Useful, if this is done anyways in the subsequent workflow
+      --index_files            [file]  Path to bai index files
+     --no_read_QC             [bool]  If specified, no quality control will be performed on extracted reads. Useful, if this is done anyways in the subsequent workflow
```
Comment (FriederikeHanssen) [inline main.nf:33]:
```suggestion
      --no_read_QC             [bool]  If specified, no quality control will be performed on extracted reads. Useful, if this is done anyways in the subsequent workflow
```

File: main.nf [hunk 14]
```diff
@@ -92,13 +94,30 @@ if(params.input && !params.chr) { //Checks whether bam file(s) and no chromosome
         .map { file -> tuple(file.name.replaceAll(".bam",''), file) } // map bam file name w/o bam to file 
         .into { bam_chr;
                 bam_files_flagstats;
+		bam_files_index;
```
Comment (FriederikeHanssen) [inline main.nf:97]:
```suggestion
		  bam_files_index;
```

File: main.nf [hunk 15]
```diff
@@ -82,6 +83,7 @@ if(params.input && !params.chr) { //Checks whether bam file(s) and no chromosome
         .map { file -> tuple(file.name.replaceAll(".bam",''), file) } // map bam file name w/o bam to file 
         .into { bam_files_check; 
                 bam_files_flagstats;
+		bam_files_index;
```
Comment (FriederikeHanssen) [inline main.nf:86]:
```suggestion
		  bam_files_index;
```

File: main.nf [hunk 16]
```diff
@@ -180,29 +200,113 @@ process get_software_versions {
...
+	if (params.chr){
+	  process extractReadsMappingToChromosome{
+	    tag "${name}.${chr_list_joined}"
+	    label 'process_medium'
   
-    input:
-    set val(name), file(bam) from bam_chr 
+	    input:
+	    set val(name), file(bam) from bam_chr 
+		  set val(name), file(bai) from ch_chr_index_files
```
Comment (ggabernet) [inline main.nf:233]:
```suggestion
	set val(name_bai), file(bai) from ch_chr_index_files
```

File: main.nf [hunk 17]
```diff
@@ -180,29 +200,113 @@ process get_software_versions {
...
+
+	    output:
+	    set val("${name}.${chr_list_joined}"), file("${name}.${chr_list_joined}.bam") into bam_files_check
+		                
+
+	    script:
+	    //If multiple chr were specified, then join space separated list for naming: chr1 chr2 -> chr1_chr2, also resolve region specification with format chr:start-end
+	    chr_list_joined = params.chr.split(' |-|:').size() > 1 ? params.chr.split(' |-|:').join('_') : params.chr
+	    """    
+	    samtools view -hb $bam ${params.chr} -@$task.cpus -o ${name}.${chr_list_joined}.bam
```
Comment (ggabernet) [inline main.nf:243]:
```suggestion
	    samtools view -hb $bam ${params.chr} -@$task.cpus -o "${name}.${chr_list_joined}.bam"
```

File: main.nf [hunk 18]
```diff
@@ -180,29 +200,113 @@ process get_software_versions {
...
+	 }
+    }
+		
+	process computeIdxstatsInput {	
+		tag "$name"
+		label 'process_medium'
+		
+		input:
+		set val(name), file(bam) from bam_files_idxstats
+		set val(name), file(bai) from ch_index_files
```
Comment (ggabernet) [inline main.nf:254]:
```suggestion
		set val(name_bai), file(bai) from ch_index_files
```

File: main.nf [hunk 19]
```diff
@@ -180,29 +200,113 @@ process get_software_versions {
...
-  }
+	// Extract reads mapping to specific chromosome(s)
+	if (params.chr){
+	  process extractReadsMappingToChromosomeBAI{
+	    tag "${name}.${chr_list_joined}"
+	    label 'process_medium'
+  
+	    input:
+	    set val(name), file(bam) from bam_chr 
+	    set val(name), file(bai) from bai_files_chr
```
Comment (ggabernet) [inline main.nf:278]:
```suggestion
	    set val(name_bai), file(bai) from bai_files_chr
```

File: main.nf [hunk 20]
```diff
@@ -180,29 +200,113 @@ process get_software_versions {
...
+		
+		input:
+		set val(name), file(bam) from bam_files_idxstats
+		set val(name), file(bai) from ch_index_files
+				
+		output:
+		file "*.idxstats" into ch_bam_idxstat_mqc		
+		script:
+		"""				
+		samtools idxstats $bam > ${bam}.idxstats
```
Comment (ggabernet) [inline main.nf:260]:
```suggestion
		samtools idxstats $bam > "${bam}.idxstats"
```

File: main.nf [hunk 21]
```diff
@@ -180,29 +200,113 @@ process get_software_versions {
...
+	    set val(name), file(bai) from bai_files_chr
+
+	    output:
+	    set val("${name}.${chr_list_joined}"), file("${name}.${chr_list_joined}.bam") into bam_files_check
+
+	    script:
+	    //If multiple chr were specified, then join space separated list for naming: chr1 chr2 -> chr1_chr2, also resolve region specification with format chr:start-end
+	    chr_list_joined = params.chr.split(' |-|:').size() > 1 ? params.chr.split(' |-|:').join('_') : params.chr
+	    """    
+	    samtools view -hb $bam ${params.chr} -@$task.cpus -o ${name}.${chr_list_joined}.bam
```
Comment (ggabernet) [inline main.nf:287]:
```suggestion
	    samtools view -hb $bam ${params.chr} -@$task.cpus -o "${name}.${chr_list_joined}.bam"
```

File: main.nf [hunk 22]
```diff
@@ -180,29 +200,113 @@ process get_software_versions {
...
+
+		input:
+		set val(name), file(bam) from bam_files_idxstats
+		set val(name), file(bai) from bai_files_idxstats
+		
+		output:
+		file "*.idxstats" into ch_bam_idxstat_mqc	
+		script:
+		"""
+		samtools idxstats $bam > ${bam}.idxstats
```
Comment (ggabernet) [inline main.nf:305]:
```suggestion
		samtools idxstats $bam > "${bam}.idxstats"
```

File: main.nf [hunk 23]
```diff
@@ -92,13 +94,30 @@ if(params.input && !params.chr) { //Checks whether bam file(s) and no chromosome
...
+
+if (params.index_files) {
+	Channel
+	.fromPath(params.index_files, checkIfExists:true)	
+	.map{file -> tuple(file.name.replaceAll(".bai", ''), file)}	
+	.into { bai_files_idxstats;
+	  	bai_files_chr}
+	
+}
+else {
```
Comment (ggabernet) [inline main.nf:117]:
```suggestion
} else {
```

File: main.nf [hunk 24]
```diff
@@ -180,29 +200,113 @@ process get_software_versions {
...
+		set val(name), file(bam) from bam_files_index		
+	
+		output:
+		file "${name}.bai" into ch_index_files 
+		file "${name}.bai" into ch_chr_index_files
+	
+		
+		script:	
+		"""		
+		samtools index $bam > ${name}.bai
```
Comment (ggabernet) [inline main.nf:221]:
```suggestion
		samtools index $bam > "${name}.bai"
```

File: main.nf [hunk 25]
```diff
@@ -180,29 +200,113 @@ process get_software_versions {
...
+    }
+	
+
+	process computeIdxstatsInputBAI {	
+		tag "$name"
+		label 'process_medium'	
+
+		input:
+		set val(name), file(bam) from bam_files_idxstats
+		set val(name), file(bai) from bai_files_idxstats
```
Comment (ggabernet) [inline main.nf:299]:
```suggestion
		set val(name_bai), file(bai) from bai_files_idxstats
```

**Review by ggabernet — COMMENTED** [review]
Hi Susanne,

thanks for your PR, I've added a couple of suggestions of what can be going wrong

File: main.nf (hunk 16 above)
Comment (ggabernet) [reply main.nf:233]: Here it is best to give different names to the two variables, even if they have the same content.

File: main.nf (hunk 17 above)
Comment (ggabernet) [reply main.nf:243]: I've seen before that not adding the `"` when having a string that contains variables inside but also additional characters can give rise to problems.

File: main.nf [hunk 26]
```diff
@@ -180,29 +199,113 @@ process get_software_versions {
...
+	if (params.chr){
+	  process extractReadsMappingToChromosome{
+	    tag "${name}.${chr_list_joined}"
+	    label 'process_medium'
   
-    input:
-    set val(name), file(bam) from bam_chr 
+	    input:
+	    set val(name), file(bam) from bam_chr 
+	set val(name_bai), file(bai) from ch_chr_index_files
```
Comment (ggabernet) [inline main.nf:232]:
```suggestion
	file(bai) from ch_chr_index_files
```

File: main.nf [hunk 27]
```diff
@@ -180,29 +199,113 @@ process get_software_versions {
...
+	 }
+    }
+		
+	process computeIdxstatsInput {	
+		tag "$name"
+		label 'process_medium'
+		
+		input:
+		set val(name), file(bam) from bam_files_idxstats
+		set val(name_bai), file(bai) from ch_index_files
```
Comment (ggabernet) [inline main.nf:253]:
```suggestion
		file(bai) from ch_index_files
```

File: main.nf [hunk 28]
```diff
@@ -180,29 +199,113 @@ process get_software_versions {
...
-  }
+	// Extract reads mapping to specific chromosome(s)
+	if (params.chr){
+	  process extractReadsMappingToChromosomeBAI{
+	    tag "${name}.${chr_list_joined}"
+	    label 'process_medium'
+  
+	    input:
+	    set val(name), file(bam) from bam_chr 
+	    set val(name_bai), file(bai) from bai_files_chr
```
Comment (ggabernet) [inline main.nf:277]:
```suggestion
	    file(bai) from bai_files_chr
```

File: main.nf [hunk 29]
```diff
@@ -180,29 +199,113 @@ process get_software_versions {
...
+    }
+	
+
+	process computeIdxstatsInputBAI {	
+		tag "$name"
+		label 'process_medium'	
+
+		input:
+		set val(name), file(bam) from bam_files_idxstats
+		set val(name_bai), file(bai) from bai_files_idxstats
```
Comment (ggabernet) [inline main.nf:298]:
```suggestion
		file(bai) from bai_files_idxstats
```

**Review by ggabernet — COMMENTED** [review]
sorry just realized the channel does not contain the names

**Review by ggabernet — COMMENTED** [review]
sorry just realized the channel does not contain the names

File: conf/test.config [hunk 30]
```diff
@@ -16,9 +16,9 @@ params {
   max_time = 48.h
   
   // Input data
-  input = ['https://github.com/nf-core/test-datasets/raw/eager/testdata/Mammoth/bam/JK2782_TGGCCGATCAACGA_L008_R1_001.fastq.gz.tengrand.fq.combined.fq.mapped.bam',
-           "$baseDir/testdata/First_SmallTest_Paired.bam",
-           "$baseDir/testdata/Second_SmallTest_Paired.bam",
-           "$baseDir/testdata/wgEncodeUwRepliSeqK562G1AlnRep1.bam"
+  input = ['https://raw.githubusercontent.com/nf-core/test-datasets/eager/testdata/Mammoth/bam/JK2782_TGGCCGATCAACGA_L008_R1_001.fastq.gz.tengrand.fq.combined.fq.mapped.bam',
+          'https://raw.githubusercontent.com/qbic-pipelines/bamtofastq/master/testdata/First_SmallTest_Paired.bam',
+          'https://raw.githubusercontent.com/qbic-pipelines/bamtofastq/master/Second_SmallTest_Paired.bam',
```
Comment (ggabernet) [inline conf/test.config:21]:
```suggestion
          'https://raw.githubusercontent.com/qbic-pipelines/bamtofastq/master/testdata/Second_SmallTest_Paired.bam',
```

File: main.nf [hunk 31]
```diff
@@ -180,29 +199,113 @@ process get_software_versions {
...
+	process IdxBAI {	
+		tag "$name"
+		label 'process_medium' 
+
+		input:
+		set val(name), file(bam) from bam_files_index		
+	
+		output:
+		file "${name}.bai" into ch_index_files 
+		file "${name}.bai" into ch_chr_index_files
```
Comment (ggabernet) [inline main.nf:215]:
```suggestion
		file "${name}.bai" into ch_chr_index_files.dump()
```

File: main.nf [hunk 32]
```diff
@@ -180,29 +199,113 @@ process get_software_versions {
...
+	
+		
+		script:	
+		"""		
+		samtools index $bam > "${name}.bai"
+		"""		
+	}
+
```
Comment (ggabernet) [inline main.nf:223]:
```suggestion
    ch_chr_index_files.dump()
```

File: main.nf [hunk 33]
```diff
@@ -180,29 +199,113 @@ process get_software_versions {
...
+	process IdxBAI {	
+		tag "$name"
+		label 'process_medium' 
+
+		input:
+		set val(name), file(bam) from bam_files_index		
+	
+		output:
+		file "${name}.bai" into ch_index_files 
+		file "${name}.bai" into ch_chr_index_files.dump()
```
Comment (ggabernet) [inline main.nf:215]:
```suggestion
		file "${name}.bai" into ch_chr_index_files
```

File: main.nf [hunk 34]
```diff
@@ -180,29 +199,113 @@ process get_software_versions {
...
+		file "${name}.bai" into ch_index_files 
+		file "${name}.bai" into ch_chr_index_files
+	
+		
+		script:	
+		"""		
+		samtools index $bam > "${name}.bai"
+		"""		
+	}
+    ch_chr_index_files.dump()
```
Comment (ggabernet) [inline main.nf:223]:
```suggestion
    ch_dump = ch_chr_index_files.dump()
```

File: main.nf [hunk 35]
```diff
@@ -180,29 +199,113 @@ process get_software_versions {
...
+	if (params.chr){
+	  process extractReadsMappingToChromosome{
+	    tag "${name}.${chr_list_joined}"
+	    label 'process_medium'
   
-    input:
-    set val(name), file(bam) from bam_chr 
+	    input:
+	    set val(name), file(bam) from bam_chr 
+	file(bai) from ch_chr_index_files
```
Comment (ggabernet) [inline main.nf:232]:
```suggestion
	file(bai) from ch_dump
```

File: main.nf (hunk 31 above)
Comment (ggabernet) [inline main.nf:215]:
```suggestion
		set val(name), file("${name}.bam"), file("${name}.bai") into(ch_bam_bai, ch_chr_bam_bai)
```

File: main.nf [hunk 36]
```diff
@@ -180,29 +199,113 @@ process get_software_versions {
...
+	 }
+    }
+		
+	process computeIdxstatsInput {	
+		tag "$name"
+		label 'process_medium'
+		
+		input:
+		set val(name), file(bam) from bam_files_idxstats
+		file(bai) from ch_index_files
```
Comment (ggabernet) [inline main.nf:253]:
```suggestion
		set val(name), file(bam), file(bai) from ch_bam_bai
```

File: main.nf [hunk 37]
```diff
@@ -180,29 +199,113 @@ process get_software_versions {
...
+	    """
+	 }
+    }
+		
+	process computeIdxstatsInput {	
+		tag "$name"
+		label 'process_medium'
+		
+		input:
+		set val(name), file(bam) from bam_files_idxstats
```
Comment (ggabernet) [inline main.nf:252]:
```suggestion
```

File: main.nf (hunk 34 above)
Comment (ggabernet) [inline main.nf]:
```suggestion
```

File: main.nf [hunk 38]
```diff
@@ -180,29 +199,113 @@ process get_software_versions {
...
+	// Extract reads mapping to specific chromosome(s)
+	if (params.chr){
+	  process extractReadsMappingToChromosome{
+	    tag "${name}.${chr_list_joined}"
+	    label 'process_medium'
   
-    input:
-    set val(name), file(bam) from bam_chr 
+	    input:
+	    set val(name), file(bam) from bam_chr
```
Comment (ggabernet) [inline main.nf:231]:
```suggestion
	    set val(name), file(bam), file(bai) from ch_chr_bam_bai
```

File: main.nf [hunk 39]
```diff
@@ -180,29 +199,113 @@ process get_software_versions {
...
+	if (params.chr){
+	  process extractReadsMappingToChromosome{
+	    tag "${name}.${chr_list_joined}"
+	    label 'process_medium'
   
-    input:
-    set val(name), file(bam) from bam_chr 
+	    input:
+	    set val(name), file(bam) from bam_chr 
+	file(bai) from ch_dump
```
Comment (ggabernet) [inline main.nf:232]:
```suggestion
```

File: main.nf [hunk 40]
```diff
@@ -180,29 +199,110 @@ process get_software_versions {
...
+		output:
+		set val(name), file("${name}.bam"), file("${name}.bai") into(ch_bam_bai, ch_chr_bam_bai)
+	
+		
+		script:	
+		"""		
+		samtools index $bam > "${name}.bai"
+		"""		
+	}
+    ch_dump = ch_chr_index_files.dump()
```
Comment (ggabernet) [inline main.nf:222]:
```suggestion
```

File: main.nf [hunk 41]
```diff
@@ -180,29 +199,109 @@ process get_software_versions {
...
+		input:
+		set val(name), file(bam) from bam_files_index		
+	
+		output:
+		set val(name), file("${name}.bam"), file("${name}.bai") into(ch_bam_bai, ch_chr_bam_bai)
+	
+		
+		script:	
+		"""		
+		samtools index $bam > "${name}.bai"
```
Comment (ggabernet) [inline main.nf:219]:
```suggestion
		samtools index $bam
```

File: main.nf [hunk 42]
```diff
@@ -180,29 +199,109 @@ process get_software_versions {
...
+		input:
+		set val(name), file(bam) from bam_files_index		
+	
+		output:
+		set val(name), file("${name}.bam"), file("${name}.bai") into(ch_bam_bai, ch_chr_bam_bai)
+	
+		
+		script:	
+		"""		
+		samtools index $bam
```
Comment (ggabernet) [inline main.nf:219]:
```suggestion
		samtools index $bam "${name}.bai"
```

**Comment by FriederikeHanssen** [comment]
Looks great @SusiJo :) Thank you very much for contributing. If I am not mistaken, there is no test yet for the `bai`, right? Would be great to add it just to be sure it works at all times. From my side we can merge now though and I add the test, when I make the stats stuff optional.

**Pull request merged**